docs: state management (#8547)
sveltejs/kit: documentation/docs/20-core-concepts/60-state-management.md

---
title: State Management
---

Managing state is one of the hardest parts in application development. In this section, we will go over the different types of state, how SvelteKit can help you with managing it, and different patterns for various use cases.

## Types of state

Apps have a multitude of different states. State can broadly be classified by two adjacent categories: The semantic category of the state, and how it is represented.

Semantic categories of state are:
- Local state: State that is only relevant inside one component, and managed inside it.
- Global state: State that is relevant to many or all parts of the application, and is likely modified from different places.

How state can be represented:
- JavaScript variable: Not visible from the outside
- URL: What page you're on, but also information for the page itself through query parameters for example

> - Cookies: Mostly modifiable through the server, contains for example information about the current user. Can be used to persist state across browser tabs or reloads.

Reviewer: ```suggestion
- [Cookies](/docs/load#cookies-and-headers): Mostly modifiable through the server. Can be used to persist state across browser tabs or reloads. May contain, for example, information about the current user.
```

Consider the following example where the user is set from inside a `load` function:

```js 
/// file: +page.js
// @filename: ambient.d.ts
declare module '$lib/user' {
	export const user: { set: (value: any) => void };
}

// @filename: index.js
// ---cut---
// DON'T DO THIS!
import { user } from '$lib/user';

/** @type {import('./$types').PageLoad} */
export async function load({ fetch }) {
    const response = await fetch('/api/user');
    user.set(await response.json());
}
```

If you are using SSR, the `load` function will run on the server initially, which means that the whole server instance which serves _all_ requests from _all_ users has its `user` state set to the one just requested from the API. To scope this to a single user, you have a couple of options:

- if you need to access the user state only inside `load` functions, use `locals` in server `load` functions

> - if you need to persist the user state across reloads, but only need to access it inside `load` functions, use `cookies` in server `load` functions

Reviewer: ```suggestion
- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use [`cookies` in server `load` functions](/docs/load#cookies-and-headers)
```

> - if you need to access and update the state inside components, use Svelte's [context feature](https://svelte.dev/docs#run-time-svelte-setcontext). That way, the state is scoped to components, which means they are not shared across different requests on the server. The drawback is that you can only set and subscribe to the store at component initialization. SvelteKit's stores from `$app/stores` for example are setup like this (which is why you may have encountered a related error message)

Reviewer 2: Could it be useful to have an example of this here ?
I am not sure myself what is meant with this, but are you referring to returning data from `load`, placing it in a store and then use `setContext` in the page to make it available ?

Author: yes this would be the way - and I agree this probably needs a code example

> ---

Reviewer 3: should the filename rather start with `50`?

title: State Management
---

Managing state is one of the hardest parts in application development. This section covers various use cases with regards to state management and what to watch out for.

> ## Careful with global state

Reviewer: ```suggestion
## Avoid global state in SSR
```

Author: Is "avoid" a too strong word, given that it's ok to have global state if you make sure to not update it to request-specific values during ssr?

Reviewer: I tend think it'd be okay as we can go into greater detail in the text and avoid just means "try not to use" and not "do not use at all"

> If you are creating an [SPA](/docs/glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](/docs/glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.

Reviewer: ```suggestion
If you are creating an [SPA](glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.
```

> If you are creating an [SPA](/docs/glossary#csr) with SvelteKit, you can create global state freely, as you can be sure that it's only initialized inside the user's browser. If you use [SSR](/docs/glossary#ssr) however, you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.

Reviewer 3: ```suggestion
If you are creating a [SPA](/docs/glossary#csr) with SvelteKit, you can create a global state freely, as you can be sure that it's only initialized inside the user's browser. However, if you use [SSR](/docs/glossary#ssr), you have to watch out for a couple of things when managing state. In many server environments, a single instance of your app will serve multiple users (this is not specific to SvelteKit - it's one of the gotchas of working with such environments). For that reason, per-request or per-user state must not be stored in global variables.
```

- if you need to access the user state only inside `load` functions, use `locals` in server `load` functions
- if you need to persist the user state across reloads, but only need to access it inside `load` functions, use `cookies` in server `load` functions
- if you need to access and update the state inside components, use Svelte's [context feature](https://svelte.dev/docs#run-time-svelte-setcontext). That way, the state is scoped to components, which means they are not shared across different requests on the server. The drawback is that you can only set and subscribe to the store at component initialization. SvelteKit's stores from `$app/stores` for example are setup like this (which is why you may have encountered a related error message)

If you have global data whose initial state is not dependent on a request (in other words, it's always the same), then you can keep storing that data globally, as long as you make sure you don't update it during the initial rendering on the server (during load or component render).

## Managing forms

> If you are coming from regular Svelte, you may be used to do something like this when creating a form to react to user input:

Reviewer: ```suggestion
If you are coming from regular Svelte, you may be used to doing something like this when creating a form to react to user input:
```
